// owoify/index.js
const MAP = {
    '+' : '*growl* ',
    '-' : '*growl* ',
    '×' : '*growl* ',
    '÷': '*growl* ',
    '/' : '*rawr?* ',
  
    '〒': '*rawr?* ',
  
    A: 'woof ',
    B: 'woofw ',
    C: 'worf ',
    D: 'worfw ',
    E: 'bark ',
    F: 'barkb ',
    G: 'arf ',
    H: 'arf arf ',
    I: 'arf arf arf ',
    J: 'bork ',
    K: 'bork bork ',
    L: '*pant* ',
    M: '*lick* ',
    N: 'arwoo ',
    O: 'ruff ',
    P: 'yaP ',
    Q: 'yap',
    R: '*rawr* ',
    S: '*howls* ',
    T: 'yap yap ',
    U: 'bow-wow ',
    V: 'bow ',
    W: 'wOw ',
    X: 'w00f',
    Y: 'r0ff',
    Z: 'bay??? ',
    a: 'w0of ',
    b: 'w00f w00f ',
    c: 'w0e0f ',
    d: 'aarf ',
    e: 'w0e0f w0e0f ',
    f: 'arfff ',
    g: '*whimpers* ',
    h: 'arfh ',
    i: 'iaarf ',
    j: 'jarf ',
    k: 'owo ',
    l: 'larf ',
    m: 'm0rk ',
    n: 'noof ',
    o: 'boof ',
    p: '*pants* ',
    q: '*w0rf* ',
    r: '*r0ff* ',
    s: 'orf ',
    t: 'b0rt ',
    u: 'orf ',
    v: 'vorf ',
    w: 'wo0o00o0of ',
    x: 'arw00o0 ',
    y: 'y0o0rf ',
    z: 'zo000orf ',
    0: '*arfs* ',
    1: '*arfs* ',
    2: '*arfs* ',
    3: '*arfs* ',
    4: '*arfs* ',
    5: '*arfs* ',
    6: '*arfs* ',
    7: '*arfs* ',
    8: '*arfs* ',
    9: '*arfs* ',
  
    ' ': ' ',
  };
  
  module.exports = function (str) {
    return str.split('').map(c => (MAP[c] || c)).join('');
  };
