// owoify-js/dist/owoify-js.esm.js
function _unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return _arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;

  for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i];

  return arr2;
}

function _createForOfIteratorHelperLoose(o, allowArrayLike) {
  var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"];
  if (it) return (it = it.call(o)).next.bind(it);

  if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") {
    if (it) o = it;
    var i = 0;
    return function () {
      if (i >= o.length) return {
        done: true
      };
      return {
        done: false,
        value: o[i++]
      };
    };
  }

  throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}

function searchValueContainsReplacedWords(searchValue, replaceValue, replacedWords) {
  return Array.from(replacedWords).some(function (word) {
    return word.replace(searchValue, replaceValue) !== word;
  });
}

var Word = /*#__PURE__*/function () {
  function Word(word) {
    this.word = word.trim();
    this.replacedWords = new Set();
  }

  var _proto = Word.prototype;

  _proto.replace = function replace(searchValue, replaceValue, replaceReplacedWords) {
    if (replaceReplacedWords === void 0) {
      replaceReplacedWords = false;
    }

    if (!replaceReplacedWords && searchValueContainsReplacedWords(searchValue, replaceValue, this.replacedWords)) return this;
    var replacingWord = this.word.replace(searchValue, replaceValue).trim();
    var matchArray = this.word.match(searchValue);
    var replacedWords = matchArray !== null ? Array.from(matchArray).map(function (x) {
      return x.replace(x, replaceValue);
    }) : [];

    if (replacingWord !== this.word) {
      for (var _iterator = _createForOfIteratorHelperLoose(replacedWords), _step; !(_step = _iterator()).done;) {
        var word = _step.value;
        this.replacedWords.add(word);
      }
      this.word = replacingWord;
    }

    return this;
  };

  _proto.toString = function toString() {
    return this.word;
  };

  return Word;
}();

/**
 * Interleaves 2 arrays
 * @internal
 * @param a first array
 * @param b second arry
 * @returns array with interleaved elements
 */
function interleaveArrays(a, b) {
  var arr = [];
  var observed = a;
  var other = b;
  var temp = null;

  while (observed.length > 0) {
    arr.push(observed.shift());
    temp = observed;
    observed = other;
    other = temp;
  }

  if (other.length > 0) arr.push.apply(arr, other);
  return arr;
}

var faces = ['(・`ω´・)', ';;w;;', 'owo', 'UwU', '>w<', '^w^', '(* ^ ω ^)', '(⌒ω⌒)', 'ヽ(*・ω・)ﾉ', '(o´∀`o)', '(o･ω･o)', '＼(＾▽＾)／'];
var Map_O_To_OwO = function Map_O_To_OwO(input) {
  return input.replace(/o/g, function () {
    return Math.round(Math.random()) ? 'owo' : 'o';
  });
};
var Map_Ew_To_UwU = function Map_Ew_To_UwU(input) {
  return input.replace(/ew/g, 'uwu');
};
var Map_Hey_To_Hay = function Map_Hey_To_Hay(input) {
  return input.replace(/([Hh])ey/g, '$1ay');
};
var Map_Dead_To_Ded = function Map_Dead_To_Ded(input) {
  return input.replace(/Dead/g, 'Ded').replace(/dead/g, 'ded');
};
var Map_N_Vowel_T_To_Nd = function Map_N_Vowel_T_To_Nd(input) {
  return input.replace(/n[aeiou]*t/g, 'nd');
};
var Map_Read_To_Wead = function Map_Read_To_Wead(input) {
  return input.replace(/Read/g, 'Wead').replace(/read/g, 'wead');
};
var Map_Brackets_To_StarTrails = function Map_Brackets_To_StarTrails(input) {
  return input.replace(/[({<]/g, '｡･:*:･ﾟ★,｡･:*:･ﾟ☆').replace(/[)}>]/g, '☆ﾟ･:*:･｡,★ﾟ･:*:･｡');
};
var Map_PeriodCommaExclamationSemicolon_To_Kaomojis = function Map_PeriodCommaExclamationSemicolon_To_Kaomojis(input) {
  return input.replace(/[.,](?![0-9])/g, function () {
    return ' ' + faces[Math.floor(Math.random() * faces.length)];
  }).replace(/[!;]+/g, function () {
    return ' ' + faces[Math.floor(Math.random() * faces.length)];
  });
};
var Map_That_To_Dat = function Map_That_To_Dat(input) {
  return input.replace(/that/g, 'dat').replace(/That/g, 'Dat');
};
var Map_Th_To_F = function Map_Th_To_F(input) {
  return input.replace(/[Tt]h(?![Ee])/g, 'f').replace(/TH(?!E)/g, 'F');
};
var Map_Le_To_Wal = function Map_Le_To_Wal(input) {
  return input.replace(/le$/g, 'wal');
};
var Map_Ve_To_We = function Map_Ve_To_We(input) {
  return input.replace(/ve/g, 'we').replace(/Ve/g, 'We');
};
var Map_Ry_To_Wwy = function Map_Ry_To_Wwy(input) {
  return input.replace(/ry/g, 'wwy');
};
var Map_ROrL_To_W = function Map_ROrL_To_W(input) {
  return input.replace(/(?:r|l)/g, 'w').replace(/(?:R|L)/g, 'W');
};
var Map_Ll_To_Ww = function Map_Ll_To_Ww(input) {
  return input.replace(/ll/g, 'ww');
};
var Map_VowelOrRExceptO_L_To_Wl = function Map_VowelOrRExceptO_L_To_Wl(input) {
  return input.replace(/[aeiur]l$/g, 'wl').replace(/[AEIUR]([lL])$/g, 'W$1');
};
var Map_Old_To_Owld = function Map_Old_To_Owld(input) {
  return input.replace(/([Oo])ld/g, '$1wld').replace(/OLD/g, 'OWLD');
};
var Map_Ol_To_Owl = function Map_Ol_To_Owl(input) {
  return input.replace(/([Oo])l/g, '$1wl').replace(/OL/g, 'OWL');
};
var Map_LOrR_O_To_Wo = function Map_LOrR_O_To_Wo(input) {
  return input.replace(/[lr]o/g, 'wo').replace(/[LR]([oO])/g, 'W$1');
};
var Map_SpecificConsonants_O_To_Letter_And_Wo = function Map_SpecificConsonants_O_To_Letter_And_Wo(input) {
  return input.replace(/([bcdfghjkmnpqstxyz])o/g, '$1wo').replace(/([BCDFGHJKMNPQSTXYZ])([oO])/g, function (_, m1, m2) {
    return m1 + (m2.toUpperCase() === m2 ? 'W' : 'w') + m2;
  });
};
var Map_VOrW_Le_To_Wal = function Map_VOrW_Le_To_Wal(input) {
  return input.replace(/[vw]le/g, 'wal');
};
var Map_Fi_To_Fwi = function Map_Fi_To_Fwi(input) {
  return input.replace(/([Ff])i/g, '$1wi').replace(/FI/g, 'FWI');
};
var Map_Ver_To_Wer = function Map_Ver_To_Wer(input) {
  return input.replace(/([Vv])er/g, 'wer');
};
var Map_Poi_To_Pwoi = function Map_Poi_To_Pwoi(input) {
  return input.replace(/([Pp])oi/g, '$1woi');
};
var Map_SpecificConsonants_Le_To_Letter_And_Wal = function Map_SpecificConsonants_Le_To_Letter_And_Wal(input) {
  return input.replace(/([DdFfGgHhJjPpQqRrSsTtXxYyZz])le$/g, '$1wal');
};
var Map_Consonant_R_To_Consonant_W = function Map_Consonant_R_To_Consonant_W(input) {
  return input.replace(/([BbCcDdFfGgKkPpQqSsTtWwXxZz])r/g, '$1w');
};
var Map_Ly_To_Wy = function Map_Ly_To_Wy(input) {
  return input.replace(/ly/g, 'wy').replace(/Ly/g, 'Wy');
};
var Map_Ple_To_Pwe = function Map_Ple_To_Pwe(input) {
  return input.replace(/([Pp])le/g, '$1we');
};
var Map_Nr_To_Nw = function Map_Nr_To_Nw(input) {
  return input.replace(/([Nn])r/g, '$1w').replace(/NR/g, 'NW');
};
var Map_Mem_To_Mwem = function Map_Mem_To_Mwem(input) {
  return input.replace(/Mem/g, 'mwem').replace(/mem/g, 'Mwem');
};
var Unmap_Nywo_To_Nyo = function Unmap_Nywo_To_Nyo(input) {
  return input.replace(/([Nn])ywo/g, '$1yo', true);
};
var Map_Fuc_To_Fwuc = function Map_Fuc_To_Fwuc(input) {
  return input.replace(/([Ff])uc/g, '$1wuc');
};
var Map_Mom_To_Mwom = function Map_Mom_To_Mwom(input) {
  return input.replace(/([Mm])om/g, '$1wom');
};
var Map_Me_To_Mwe = function Map_Me_To_Mwe(input) {
  return input.replace(/^Me$/g, 'Mwe').replace(/^me$/g, 'mwe');
};
var Map_NVowel_To_Ny = function Map_NVowel_To_Ny(input) {
  return input.replace(/n([aeiou])/g, 'ny$1').replace(/N([aeiou])/g, 'Ny$1').replace(/N([AEIOU])/g, 'NY$1');
};
var Map_Ove_To_Uv = function Map_Ove_To_Uv(input) {
  return input.replace(/ove/g, 'uv').replace(/OVE/g, 'UV');
};
var Map_Haha_To_HehexD = function Map_Haha_To_HehexD(input) {
  return input.replace(/\b(ha|hah|heh|hehe)+\b/g, 'hehe xD');
};
var Map_The_To_Teh = function Map_The_To_Teh(input) {
  return input.replace(/\b([Tt])he\b/g, '$1eh');
};
var Map_You_To_U = function Map_You_To_U(input) {
  return input.replace(/\bYou\b/g, 'U').replace(/\byou\b/g, 'u');
};
var Map_Time_To_Tim = function Map_Time_To_Tim(input) {
  return input.replace(/\b([Tt])ime\b/g, '$1im');
};
var Map_Over_To_Owor = function Map_Over_To_Owor(input) {
  return input.replace(/([Oo])ver/g, '$1wor');
};
var Map_Worse_To_Wose = function Map_Worse_To_Wose(input) {
  return input.replace(/([Ww])orse/g, '$1ose');
};
var Map_Great_To_Gwate = function Map_Great_To_Gwate(input) {
  return input.replace(/([Gg])reat/g, '$1wate');
};
var Map_Aviat_To_Awiat = function Map_Aviat_To_Awiat(input) {
  return input.replace(/([Aa])viat/g, '$1wiat');
};
var Map_Dedicat_To_Deditat = function Map_Dedicat_To_Deditat(input) {
  return input.replace(/([Dd])edicat/g, '$1editat');
};
var Map_Remember_To_Rember = function Map_Remember_To_Rember(input) {
  return input.replace(/([Rr])emember/g, '$1ember');
};
var Map_When_To_Wen = function Map_When_To_Wen(input) {
  return input.replace(/([Ww])hen/g, '$1en');
};
var Map_Frightened_To_Frigten = function Map_Frightened_To_Frigten(input) {
  return input.replace(/([Ff])righten(ed)*/g, '$1rigten');
};
var Map_Meme_To_Mem = function Map_Meme_To_Mem(input) {
  return input.replace(/Meme/g, 'mem').replace(/Mem/g, 'Mem');
};
var Map_Feel_To_Fell = function Map_Feel_To_Fell(input) {
  return input.replace(/^([Ff])eel$/g, '$1ell');
};

var specificWordMappingArray = [Map_Fuc_To_Fwuc, Map_Mom_To_Mwom, Map_Time_To_Tim, Map_Me_To_Mwe, Map_Over_To_Owor, Map_Ove_To_Uv, Map_Haha_To_HehexD, Map_The_To_Teh, Map_You_To_U, Map_Read_To_Wead, Map_Worse_To_Wose, Map_Great_To_Gwate, Map_Aviat_To_Awiat, Map_Dedicat_To_Deditat, Map_Remember_To_Rember, Map_When_To_Wen, Map_Frightened_To_Frigten, Map_Meme_To_Mem, Map_Feel_To_Fell];
var uvuMappingArray = [Map_O_To_OwO, Map_Ew_To_UwU, Map_Hey_To_Hay, Map_Dead_To_Ded, Map_N_Vowel_T_To_Nd];
var uwuMappingArray = [Map_Brackets_To_StarTrails, Map_PeriodCommaExclamationSemicolon_To_Kaomojis, Map_That_To_Dat, Map_Th_To_F, Map_Le_To_Wal, Map_Ve_To_We, Map_Ry_To_Wwy, Map_ROrL_To_W];
var owoMappingArray = [Map_NVowel_To_Ny, Map_Ll_To_Ww, Map_VowelOrRExceptO_L_To_Wl, Map_Old_To_Owld, Map_Ol_To_Owl, Map_LOrR_O_To_Wo, Map_SpecificConsonants_O_To_Letter_And_Wo, Map_VOrW_Le_To_Wal, Map_Fi_To_Fwi, Map_Ver_To_Wer, Map_Poi_To_Pwoi, Map_SpecificConsonants_Le_To_Letter_And_Wal, Map_Consonant_R_To_Consonant_W, Map_Ly_To_Wy, Map_Ple_To_Pwe, Map_Nr_To_Nw, Map_Mem_To_Mwem, Unmap_Nywo_To_Nyo];

/**
 * owoify everything
 * @param v word to owoify
 * @param level level of owo-ness. (owo -> uwu -> uvu)
 */

function owoify(v, level) {
  if (level === void 0) {
    level = 'owo';
  }

  var whitespace = v.split(/[^\s]+/g);
  var words = v.split(/\s+/g).map(function (x) {
    return new Word(x);
  });
  words = words.map(function (x) {
    specificWordMappingArray.reduce(function (word, f) {
      return f(word);
    }, x);

    switch (level) {
      case 'uvu':
        uvuMappingArray.reduce(function (word, f) {
          return f(word);
        }, x);

      case 'uwu':
        uwuMappingArray.reduce(function (word, f) {
          return f(word);
        }, x);

      case 'owo':
        owoMappingArray.reduce(function (word, f) {
          return f(word);
        }, x);
    }

    return x;
  });
  return interleaveArrays(whitespace, words).join('');
}
function uwuify(v) {
  return owoify(v, 'uwu');
}
function uvuify(v) {
  return owoify(v, 'uvu');
}

export default owoify;
export { uvuify, uwuify };
